### PR TITLE
[CppExtension Cuda] Add cuda unit test for CppExtension

### PR DESCRIPTION
**Pull request by jiahy0825**
### PR types
Others

### PR changes
Others

### Description

Pcard-66987

[CppExtension Cuda] Add cuda unit test for CppExtension.

CppExtension supports cuda codes, add a unit test to monitor this feature.

Finish one TODO item in https://github.com/PaddlePaddle/Paddle/pull/49893.

Update: Patch codes to pybind/cast.h, because pybind and the cuda12's nvcc compiler are incompatible. Detailed PRs are listed below.

由于 cuda12 的 nvcc 编译器和 pybind 有冲突，因此需要往 pybind 代码里打补丁，并添加了 Extension 机制支持 CUDA 的单测，

### CODE REVIEW

**Comment by jiahy0825** [comment]
Error raises when using CUDA 12 and pybind11, see:
- https://github.com/pybind/pybind11/issues/4606
- https://github.com/JamesPerlman/TurboNeRF/pull/23
CI link: https://xly.bce.baidu.com/paddlepaddle/paddle/newipipe/detail/8240380/job/22452437

We also need to patch codes.

**Review by zyfncg — APPROVED** [review]

File: cmake/external/pybind11.cmake [hunk 1]
```diff
@@ -33,6 +41,7 @@ ExternalProject_Add(
   # third-party library version changes cannot be incorporated.
   # reference: https://cmake.org/cmake/help/latest/module/ExternalProject.html
   UPDATE_COMMAND ""
+  PATCH_COMMAND ${PYBIND_PATCH_COMMAND}
```
Comment (YuanRisheng) [inline cmake/external/pybind11.cmake:44]: 这个是不是只能作为临时方案，后边如果pybind第三方代码有变动是不是会有问题，如果是临时方案的话可以加个TODO

**Review by YuanRisheng — APPROVED** [review]

File: cmake/external/pybind11.cmake [hunk 2]
```diff
@@ -21,6 +21,14 @@ set(PYBIND_TAG v2.10.3)
 set(PYBIND_INCLUDE_DIR ${THIRD_PARTY_PATH}/pybind/src/extern_pybind/include)
 include_directories(${PYBIND_INCLUDE_DIR})
 
+set(PYBIND_PATCH_COMMAND "")
+if(NOT WIN32)
+  file(TO_NATIVE_PATH ${PADDLE_SOURCE_DIR}/patches/pybind/cast.h.patch
```
Comment (YuanRisheng) [inline cmake/external/pybind11.cmake:26]: 这里为啥要变换一下路径，直接使用`${PADDLE_SOURCE_DIR}/patches/pybind/cast.h.patch`不可以吗

**Review by heavyrain-lzy — APPROVED** [review]
LGTM

File: cmake/external/pybind11.cmake (hunk 1 above)
Comment (jiahy0825) [reply cmake/external/pybind11.cmake:44]: pybind 现在已经固定 tag 了，所以 pybind 代码更新不会引入问题，后续 pybind 再有升级 tag 的计划时，可以考虑去掉 patch

Paddle has already fixed the tag version of pybind, so no problem would be involved if pybind updates it's codes.
We can remove this patch if Paddle updates pybind to a new tag version.

File: cmake/external/pybind11.cmake (hunk 2 above)
Comment (jiahy0825) [reply cmake/external/pybind11.cmake:26]: 也可以直接使用路径，此处是为了和其他 cmake 的 PATCH 操作保持一致

We can also use cmake path, use `TO_NATIVE_PATH` here to be consistent with other cmake files in Paddle.

Ref: https://www.jianshu.com/p/ed151fdcf473

image

**Pull request merged**